### PR TITLE
Add GH Issue Scores

### PR DESCRIPTION
**Pull request by hubertdeng123**
This will enable sending weekly GH engagement scores to slack by querying BigQuery.

Screenshot 2023-10-12 at 3 46 54 PM

A new google cron job will be set up in the ops repo here: https://github.com/getsentry/ops/pull/8194

Initially, this will send to the team-ospo channel, which we'll change to the eng management channel. A part of https://github.com/getsentry/team-ospo/issues/176

### CODE REVIEW

File: src/utils/scores.ts
Comment (hubertdeng123) [inline src/utils/scores.ts:7]: Don't see a good way to test this specifically 😔

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:34]: Add space and remove redundant `team-` prefix?

```suggestion
| ospo                               | 0/0 (100%)               |
```

Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts]: done!

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:97]:
```suggestion
| Team                          │ % on Time |
```

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:89]:
```suggestion
            text: '🗓️ Weekly GitHub Response Times by Team 🗓️',
```

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:99]:
```suggestion
| ospo                          | 100 (1/1)               |
```

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:101]: Push n=0 to bottom?

Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts]: done

**Comment by hubertdeng123** [comment]
hmmm lemme add padding on the percentage number

**Comment by hubertdeng123** [comment]
better!
Screenshot 2023-10-12 at 3 16 53 PM

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:102]:
```suggestion
| ingest                        |   - (0/0)      |
```

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [inline src/webhooks/pubsub/slackScores.test.ts:103]: "null" is weird ... what does it mean? Would "other" or "no team" be clearer?

Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts:103]: It's the name I'm using to indicate that the `slack_channel` and `offices` fields in product-owners.yml is null. I don't think that other or no team would be clearer for that.

File: src/webhooks/pubsub/slackScores.test.ts
Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts:102]: good idea

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [reply src/webhooks/pubsub/slackScores.test.ts]: Oh, hrm ... well I clearly didn't understand what it means! 😅

We can only know what team an issue is for if we have a `slack_channel` and `offices`?

Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts]: Not quite, it's a team I've used for other test suites. It's something that's useful to test to ensure eng-pipes doesn't break when slack notifications are being sent out to a team without a `slack_channel` or `offices`, but it's not useful for this PR at all. It's loaded in to the test suite on initialization which is why it appeared here. Does that make sense?

It's defined [here](https://github.com/getsentry/eng-pipes/blob/main/test/product-owners.yml#L42)

I have attempted to try and reload `product-owners.yml` for each test case with no luck and eventually gave up, so it's shared between test files right now 😔

File: src/utils/scores.ts
Comment (azaslavsky) [reply src/utils/scores.ts:7]: You mean mocking out bigquery? Could we wrap it in an interface that only has methods for getting the stuff we need from BigQuery, then mock that out?

Comment (azaslavsky) [reply src/utils/scores.ts:7]: Another idea is just to create a spy that verifies that the exact SQL we expect gets sent, and provides mock results.

File: src/webhooks/pubsub/slackScores.test.ts
Comment (azaslavsky) [inline src/webhooks/pubsub/slackScores.test.ts:31]: nit: I think we should have a leading newline in the output (maybe a trailing one too?), if for no other reason than that it makes tests easier to read ;)

File: src/webhooks/pubsub/slackScores.test.ts
Comment (azaslavsky) [inline src/webhooks/pubsub/slackScores.test.ts:34]: nit: results that are equivalent in all other sorting metric (just % on time, right?) should be sorted alphabetically. This will make tests more deterministic, and the message itself easier to parse for users.

File: src/webhooks/pubsub/slackScores.test.ts
Comment (azaslavsky) [inline src/webhooks/pubsub/slackScores.test.ts:99]: What if there's another team with the same percentage, but a different number of issues (ex: team foobar has 2/2, and team bazqux has 2/4)? Proposal: we should order first by % complete descending, then by number of issues descending, then by alphabetical order ascending. So, we would have:

Frank | 100% (2/2)
Grace | 100% (1/1)
Alice | 50% (2/4)
Charlie | 50% (1/2)
Boris | 0% (0/0)
Daniella | 0% (0/0)

Wdyt?

File: src/webhooks/pubsub/slackScores.ts
Comment (azaslavsky) [inline src/webhooks/pubsub/slackScores.ts:53]: Should the border lines be scaled to the ..._WIDTH constants from above?

File: src/webhooks/pubsub/slackScores.ts
Comment (azaslavsky) [inline src/webhooks/pubsub/slackScores.ts:31]: Why is this -1, and not 0? Am I correct in assuming that it's to represent the different between "team had no events" vs "team had events, but did 0 of them". If so, I'd do an optional type instead here, that way the type system will help you ensure that all edge cases and combinations are covered. Otherwise, it's very easy write a bug where we forget to check the -1 case and end up with funky math (made especially dangerous by JS/TSs lack of a uint type!).

**Review by azaslavsky — CHANGES_REQUESTED** [review]

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [reply src/webhooks/pubsub/slackScores.test.ts:99]: I'm with @azaslavsky. In fact, I thought I said this already but maybe not. :o)

File: src/webhooks/pubsub/slackScores.test.ts
Comment (chadwhitacre) [reply src/webhooks/pubsub/slackScores.test.ts]: Ah! So `null` only shows up in test suites and not in prod?

Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts]: yep, it only appears in test suites

Comment (chadwhitacre) [reply src/webhooks/pubsub/slackScores.test.ts]: Got it, thanks!

**Comment by chadwhitacre** [comment]
Do we want a link to ... somewhere? So people can drill down and potentially take action? Maybe like this?

> [Looker Dashboard](https://sentryio.cloud.looker.com/dashboards/401) ❧ [Project Board](https://github.com/orgs/getsentry/projects/76)

File: src/webhooks/pubsub/slackScores.ts
Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.ts:31]: It was -1 instead of 0 since teams with 0/0 issues should be sorted below those with 0/1. However, with the idea where we should sort by number of issues each team has received descending, that idea seems to be better here. I'll change this.

File: src/webhooks/pubsub/slackScores.test.ts
Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts:99]: good idea

Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.test.ts:101]: pushed!

File: src/utils/scores.ts
Comment (hubertdeng123) [reply src/utils/scores.ts:7]: Ended up just creating a spy that verifies the exact sql

**Comment by hubertdeng123** [comment]
> Do we want a link to ... somewhere? So people can drill down and potentially take action? Maybe like this?

I thought about this. At the moment there isn't a good view where people can see which responses are not triaged on time. Teams may have relatively bad `% on time` but nothing left to triage, so there isn't a direct correlation there. Although I guess it is helpful to get teams to still see which issues are still not triaged so they can address them. Guess there's no harm in adding

**Comment by chadwhitacre** [comment]
> I thought about this.

Fair point. I guess the question is, what do we expect people to do with this report?

**Comment by hubertdeng123** [comment]
> Fair point. I guess the question is, what do we expect people to do with this report?

Mostly to increase visibility and to inform managers how well their team is responding to issues. Teams already have a way of tackling issues that are overdue through slack triage notifications

File: src/utils/scores.test.ts
Comment (azaslavsky) [inline src/utils/scores.test.ts:13]: Just to be clear: this mock applies to all test cases, not just the ones in this file? In other words, is there anywhere in these tests that we're testing against a live BigQuery instance?

File: src/webhooks/pubsub/slackScores.ts
Comment (azaslavsky) [inline src/webhooks/pubsub/slackScores.ts]: nit: Still think this is much clearer as an Optional type, rather than a special number you have to check against (ex: not obvious from the type that -1 is the only special number). But not going to block the PR on it, either way works.

**Review by azaslavsky — APPROVED** [review]
LGTM assuming we're not calling to BigQuery's live API over the network in any tests.

File: src/utils/scores.test.ts
Comment (hubertdeng123) [reply src/utils/scores.test.ts:13]: None of the tests are against a live BigQuery instance

File: src/webhooks/pubsub/slackScores.ts
Comment (hubertdeng123) [reply src/webhooks/pubsub/slackScores.ts]: The reason I didn't change it to be an Optional type is because it introduces some more complex logic to the sorting function that is used later. Otherwise, there needs to be some sort of conditional in the sorting logic specifically for when the variable is undefined. While, if we have a special number 0 for the 0/0 case, the team will be placed at the bottom of the scoreboard automatically. I will write a comment here so that it's more clear.

**Pull request merged**
